activerecord/lib/active_record/associations/association_scope.rb

```diff
@@ -134,13 +134,13 @@ def add_constraints(scope, owner, association_klass, refl, chain_head, chain_tai
             scope = next_chain_scope(scope, table, reflection, association_klass, foreign_table, next_reflection)
           end
 
-          # Exclude the scope of the association itself, because that
-          # was already merged in the #scope method.
           reflection.constraints.each do |scope_chain_item|
-            item  = eval_scope(reflection.klass, scope_chain_item, owner)
+            item = eval_scope(reflection.klass, scope_chain_item, owner)
 
             if scope_chain_item == refl.scope
               scope.merge! item.except(:where, :includes)
+            elsif item.joins_values.any?
+              merge_joins(scope, item)
```

Comment:
That doesn't appear to be the case. The code is almost exactly identical to what you have here. https://github.com/rails/rails/blob/master/activerecord/lib/active_record/relation/merger.rb#L107-L129